events/source/equeue.c

```diff
@@ -21,6 +21,12 @@
 #include <stdint.h>
 #include <string.h>
 
+// check if the event is allocaded by user - event address is outside queues internal buffer address range
+#define IS_USER_ALLOCATED_EVENT(e) (((unsigned int)(e) < (unsigned int)q->buffer) || ((unsigned int)(e) > ((unsigned int)q->slab.data)))
+// for user allocated events set id as its address with first bit set
+#define MAKE_USER_ALLOCATED_EVENT_ID(e) ((int)(((unsigned int)e) | 1))
```

Comment:
Okay, this is potentially a bit icky when we need to squeeze a pointer into the `int`,  We probably get away with it for now, but at some point might need to think about making the `int` be `intptr_t`, or actually storing an ID independently inside the object rather than mapping.

Reply:
or extend `equeue` API by `void equeue_cancel_by_ptr(equeue_t *queue, struct equeue_event *e);` ?

Reply:
Oh, yes, indeed. No point using ID numbers with user-allocated events.

Reply:
But this force changing `UserAllocatedEvent` API to not provide event id to outside world.
Only `UserAllocatedEvent ::cancel` will work and  `EventQueue::cancel` will not.
I'm not sure how similar the `Event` and `UserAllocatedEvent` APIs should be?

Reply:
`EventQueue::cancel` could have an overload accepting a `UserAllocatedEvent *`.

I don't think there's any fundamental reason for them to mirror each other that degree. For a user-allocated event, an ID isn't fundamentally useful. And the post cannot fail anyway, so making any posts be `void`-returning enforces that more than just saying "this won't return 0".